File: studyq-backend/routes/auth_routes.py
from flask import Blueprint, request
from flask_jwt_extended import jwt_required, get_jwt_identity

from models import db
from models.user import User
from utils.helpers import success_response, error_response, validate_email, validate_password
from utils.jwt_utils import generate_token

auth_bp = Blueprint("auth", __name__)


@auth_bp.route("/register", methods=["POST"])
def register():
    data = request.get_json(silent=True) or {}
    username = (data.get("username") or "").strip()
    email = (data.get("email") or "").strip().lower()
    password = data.get("password") or ""

    # Validation
    errors = {}
    if not username or len(username) < 3:
        errors["username"] = "Username must be at least 3 characters."
    if not validate_email(email):
        errors["email"] = "Invalid email address."
    valid_pwd, pwd_msg = validate_password(password)
    if not valid_pwd:
        errors["password"] = pwd_msg
    if errors:
        return error_response("Validation failed.", 422, errors)

    # Uniqueness check
    if User.query.filter_by(email=email).first():
        return error_response("Email already registered.", 409)
    if User.query.filter_by(username=username).first():
        return error_response("Username already taken.", 409)

    user = User(username=username, email=email)
    user.set_password(password)
    db.session.add(user)
    db.session.commit()

    token = generate_token(user.id)
    return success_response(
        {"user": user.to_dict(), "token": token},
        "Registration successful.",
        201,
    )


@auth_bp.route("/login", methods=["POST"])
def login():
    data = request.get_json(silent=True) or {}
    email = (data.get("email") or "").strip().lower()
    password = data.get("password") or ""

    if not email or not password:
        return error_response("Email and password are required.", 400)

    user = User.query.filter_by(email=email).first()
    if not user or not user.check_password(password):
        return error_response("Invalid email or password.", 401)

    token = generate_token(user.id)
    return success_response(
        {"user": user.to_dict(), "token": token},
        "Login successful.",
    )


@auth_bp.route("/me", methods=["GET"])
@jwt_required()
def me():
    user_id = int(get_jwt_identity())
    user = User.query.get(user_id)
    if not user:
        return error_response("User not found.", 404)
    return success_response(user.to_dict())

File: studyq-backend/routes/authRoutes.js
const express = require('express');
const jwt = require('jsonwebtoken');
const User = require('../models/User');
const { successResponse, errorResponse, validateEmail, validatePassword } = require('../utils/helpers');
const authMiddleware = require('../middleware/auth');

const router = express.Router();

const generateToken = (userId) => {
  return jwt.sign({ sub: userId }, process.env.JWT_SECRET_KEY, { expiresIn: '7d' });
};

router.post('/register', async (req, res) => {
  try {
    const username = (req.body.username || '').trim();
    const email = (req.body.email || '').trim().toLowerCase();
    const password = req.body.password || '';

    // Validation
    const errors = {};
    if (!username || username.length < 3) {
      errors.username = "Username must be at least 3 characters.";
    }
    if (!validateEmail(email)) {
      errors.email = "Invalid email address.";
    }
    const pwdCheck = validatePassword(password);
    if (!pwdCheck.valid) {
      errors.password = pwdCheck.msg;
    }

    if (Object.keys(errors).length > 0) {
      return errorResponse(res, "Validation failed.", 422, errors);
    }

    // Uniqueness check
    const existingEmail = await User.findOne({ email });
    if (existingEmail) {
      return errorResponse(res, "Email already registered.", 409);
    }
    const existingUser = await User.findOne({ username });
    if (existingUser) {
      return errorResponse(res, "Username already taken.", 409);
    }

    const user = new User({ username, email });
    await user.setPassword(password);
    await user.save();

    const token = generateToken(user._id);
    return successResponse(res, { user: user.toDict(), token }, "Registration successful.", 201);
  } catch (error) {
    console.error(error);
    return errorResponse(res, "Server error during registration.", 500);
  }
});

router.post('/login', async (req, res) => {
  try {
    const email = (req.body.email || '').trim().toLowerCase();
    const password = req.body.password || '';

    if (!email || !password) {
      return errorResponse(res, "Email and password are required.", 400);
    }

    const user = await User.findOne({ email });
    if (!user) {
      return errorResponse(res, "Invalid email or password.", 401);
    }

    const isMatch = await user.checkPassword(password);
    if (!isMatch) {
      return errorResponse(res, "Invalid email or password.", 401);
    }

    const token = generateToken(user._id);
    return successResponse(res, { user: user.toDict(), token }, "Login successful.", 200);
  } catch (error) {
    console.error(error);
    return errorResponse(res, "Server error during login.", 500);
  }
});

router.get('/me', authMiddleware, async (req, res) => {
  try {
    const user = await User.findById(req.user.id);
    if (!user) {
      return errorResponse(res, "User not found.", 404);
    }
    return successResponse(res, user.toDict());
  } catch (error) {
    console.error(error);
    return errorResponse(res, "Server error.", 500);
  }
});

module.exports = router;
